raft/src/main/java/org/apache/kafka/raft/internals/AddVoterHandler.java

```diff
@@ -0,0 +1,383 @@
+
+        // Check that the new voter is caught up to the LEO to avoid delays in HWM increases
+        if (!leaderState.isReplicaCaughtUp(current.voterKey(), currentTimeMs)) {
+            logger.info(
+                "Aborting add voter operation for {} at {} since it is lagging behind: {}",
+                current.voterKey(),
+                current.voterEndpoints(),
+                leaderState.getReplicaState(current.voterKey())
+            );
+
+            abortAddVoter(
+                leaderState,
+                Errors.REQUEST_TIMED_OUT,
+                String.format(
+                    "Aborted add voter operation for %s since it is lagging behind",
+                    current.voterKey()
+                )
+            );
+
+            return true;
+        }
+
+        // Add the new voter to the set of voters and append the record to the log
+        VoterSet newVoters = partitionState
+            .lastVoterSet()
+            .addVoter(
+                VoterSet.VoterNode.of(
+                    current.voterKey(),
+                    current.voterEndpoints(),
+                    new SupportedVersionRange(
+                        supportedKraftVersions.get().minVersion(),
+                        supportedKraftVersions.get().maxVersion()
+                    )
+                )
+            )
+            .orElseThrow(() ->
+                new IllegalStateException(
+                    String.format(
+                        "Unable to add %s to the set of voters %s",
+                        current.voterKey(),
+                        partitionState.lastVoterSet()
+                    )
+                )
+            );
+        current.setLastOffset(leaderState.appendVotersRecord(newVoters, currentTimeMs));
+
+        return true;
+    }
+
+    public void highWatermarkUpdated(LeaderState<?> leaderState) {
+        leaderState.addVoterHandlerState().ifPresent(current -> {
+            leaderState.highWatermark().ifPresent(highWatermark -> {
+                current.lastOffset().ifPresent(lastOffset -> {
+                    if (highWatermark.offset() > lastOffset) {
+                        // VotersRecord with the added voter was committed; complete the RPC
+                        completeAddVoter(leaderState);
+                    }
+                });
+            });
+        });
+    }
+
+    public long maybeExpirePendingOperation(LeaderState<?> leaderState, long currentTimeMs) {
+        long timeUntilOperationExpiration = leaderState
+            .addVoterHandlerState()
+            .map(state -> state.timeUntilOperationExpiration(currentTimeMs))
+            .orElse(Long.MAX_VALUE);
+
+        if (timeUntilOperationExpiration == 0) {
+            abortAddVoter(leaderState, Errors.REQUEST_TIMED_OUT, null);
+            return Long.MAX_VALUE;
+        } else {
+            return timeUntilOperationExpiration;
+        }
+
+    }
+
+    private ApiVersionsRequestData buildApiVersionsRequest() {
```

Comment:
should we just use `ApiVersionsRequest.Builder` here? otherwise if we want to add new fields to `ApiVersionsRequest`, we have to add them in two places....

Reply:
The problem with the builder is that it requires a `version`. At this layer we don't know the versions supported by the destination since it is possible that the leader never created a connection, etc. Just like other requests sent by `KafkaRaftClient` it is better to let `NetworkClient` negotiate the version.

One solution is to make `ApiVersionRequest.Builder.DATA` public and AddVoterHandler just sends that. What do you think?

Reply:
There is a no-argument version of the builder which does not mandate a specific version, which you could use. I don't feel strongly, though (won't hold up the PR for this)

Reply:
I see. The no parameter version assumes the latest version. I guess this is good enough since we drop the version when sending the request to the `NetworkClient`.